File: PAL/Darwin/HAPPlatformTCPStreamManager.m
// Copyright (c) 2015-2019 The HomeKit ADK Contributors
//
// Licensed under the Apache License, Version 2.0 (the “License”);
// you may not use this file except in compliance with the License.
// See [CONTRIBUTORS.md] for the list of HomeKit ADK project authors.

#import <Foundation/Foundation.h>
#import <Network/Network.h>

#include "HAPPlatformTCPStreamManager+Init.h"

static const HAPLogObject tcp_log = { .subsystem = kHAPPlatform_LogSubsystem, .category = "TCPStreamManager" };

@interface Connection : NSObject
@property (nonatomic, retain) nw_connection_t socket;
@property nw_connection_state_t state;
@property HAPPlatformTCPStreamManagerRef _Nonnull tcpStreamManager;
@property HAPPlatformTCPStreamEvent interests;
@property HAPPlatformTCPStreamEventCallback _Nullable callback;
@property void* _Nullable context;
@property (nonatomic, retain) NSData* buffer;
@end

@implementation Connection
- (instancetype)initWithSocket:(nw_connection_t)socket manager:(HAPPlatformTCPStreamManagerRef)manager {
    if (self = [super init]) {
        _socket = socket;
        _state = nw_connection_state_invalid;
        _tcpStreamManager = manager;
        _interests = (HAPPlatformTCPStreamEvent) {
            .hasBytesAvailable = false,
            .hasSpaceAvailable = false,
        };
        _callback = NULL;
        _context = NULL;
        _buffer = [[NSMutableData alloc] init];
    }
    return self;
}
@end

static nw_listener_t listener = nil;
static NSMutableArray<Connection*>* socketsWaitingToBeAccepted = nil;
static NSMutableArray<Connection*>* connections = nil;

void HAPPlatformTCPStreamManagerCreate(
        HAPPlatformTCPStreamManagerRef _Nonnull tcpStreamManager,
        const HAPPlatformTCPStreamManagerOptions* options) {
    socketsWaitingToBeAccepted = [[NSMutableArray alloc] init];
    connections = [[NSMutableArray alloc] init];
}

void HAPPlatformTCPStreamManagerRelease(HAPPlatformTCPStreamManagerRef _Nonnull tcpStreamManager) {
    if (listener) {
        HAPPlatformTCPStreamManagerCloseListener(tcpStreamManager);
    }
    socketsWaitingToBeAccepted = nil;
    connections = nil;
}

HAP_RESULT_USE_CHECK
uint16_t HAPPlatformTCPStreamManagerGetListenerPort(HAPPlatformTCPStreamManagerRef _Nonnull tcpStreamManager) {
    HAPPrecondition(tcpStreamManager);
    HAPPrecondition(listener);

    return nw_listener_get_port(listener);
}

HAP_RESULT_USE_CHECK
bool HAPPlatformTCPStreamManagerIsListenerOpen(HAPPlatformTCPStreamManagerRef _Nonnull tcpStreamManager) {
    HAPPrecondition(tcpStreamManager);
    return listener != nil;
}

void HAPPlatformTCPStreamManagerOpenListener(
        HAPPlatformTCPStreamManagerRef _Nonnull tcpStreamManager,
        HAPPlatformTCPStreamListenerCallback _Nonnull callback,
        void* _Nullable context) {
    HAPPrecondition(tcpStreamManager);
    HAPPrecondition(callback);
    HAPPrecondition(!listener);
    HAPPrecondition(socketsWaitingToBeAccepted);

    nw_endpoint_t host = nw_endpoint_create_host("::", "0");
    nw_parameters_t tcp =
            nw_parameters_create_secure_tcp(NW_PARAMETERS_DISABLE_PROTOCOL, NW_PARAMETERS_DEFAULT_CONFIGURATION);
    nw_parameters_set_local_endpoint(tcp, host);
    listener = nw_listener_create(tcp);
    HAPAssert(listener);
    nw_listener_set_queue(listener, dispatch_get_main_queue());
    nw_listener_set_new_connection_handler(listener, ^(nw_connection_t socket) {
        Connection* connection = [[Connection alloc] initWithSocket:socket manager:tcpStreamManager];
        [socketsWaitingToBeAccepted addObject:connection];
        callback(tcpStreamManager, context);
    });
    nw_listener_start(listener);
}

void HAPPlatformTCPStreamManagerCloseListener(HAPPlatformTCPStreamManagerRef _Nonnull tcpStreamManager) {
    HAPPrecondition(tcpStreamManager);
    HAPPrecondition(listener);

    nw_listener_cancel(listener);
    listener = nil;
}

static void EventCallback(Connection* connection, bool incoming) {
    if (![connections containsObject:connection]) {
        return;
    }

    if ((connection.interests.hasSpaceAvailable && !incoming) || (connection.interests.hasBytesAvailable && incoming)) {
        connection.callback(
                connection.tcpStreamManager,
                (HAPPlatformTCPStreamRef)(__bridge void*) connection,
                (HAPPlatformTCPStreamEvent) {
                        .hasBytesAvailable = incoming,
                        .hasSpaceAvailable = !incoming,
                },
                connection.context);
    }
}

static void WaitForMoreDataInBackground(Connection* connection) {
    if (![connections containsObject:connection]) {
        return;
    }
    nw_connection_receive(
            connection.socket,
            1,
            4096,
            ^(dispatch_data_t data, nw_content_context_t context, bool is_complete, nw_error_t error) {
                if (error) {
                    return;
                }
                NSMutableData* newdata = [[NSMutableData alloc] init];
                [newdata appendData:connection.buffer];
                [newdata appendData:(NSData*) data];
                connection.buffer = newdata;
                EventCallback(connection, true);
                dispatch_async(dispatch_get_main_queue(), ^{
                    WaitForMoreDataInBackground(connection);
                });
            });
}

HAP_RESULT_USE_CHECK
HAPError HAPPlatformTCPStreamManagerAcceptTCPStream(
        HAPPlatformTCPStreamManagerRef _Nonnull tcpStreamManager,
        HAPPlatformTCPStreamRef* _Nonnull stream) {
    HAPPrecondition(tcpStreamManager);
    HAPPrecondition(stream);
    HAPPrecondition(socketsWaitingToBeAccepted.count > 0);

    Connection* connection = socketsWaitingToBeAccepted.lastObject;
    [socketsWaitingToBeAccepted removeLastObject];
    HAPAssert(![connections containsObject:connection]);
    [connections addObject:connection];
    *stream = (HAPPlatformTCPStreamRef)(__bridge void*) connection;

    nw_connection_set_queue(connection.socket, dispatch_get_main_queue());
    nw_connection_set_state_changed_handler(connection.socket, ^(nw_connection_state_t state, nw_error_t error) {
        connection.state = state;
        if (error) {
            return;
        }
        if (state == nw_connection_state_ready) {
            EventCallback(connection, false);
        }
    });
    WaitForMoreDataInBackground(connection);
    nw_connection_start(connection.socket);

    HAPLogDebug(&tcp_log, "Accept %lx", *stream);

    return kHAPError_None;
}

void HAPPlatformTCPStreamCloseOutput(
        HAPPlatformTCPStreamManagerRef _Nonnull tcpStreamManager,
        HAPPlatformTCPStreamRef tcpStream) {
    HAPPrecondition(tcpStreamManager);
    HAPPrecondition(tcpStream);

    HAPLogDebug(&tcp_log, "CloseOutput %lx", tcpStream);

    Connection* connection = (__bridge Connection*) (void*) tcpStream;
    HAPAssert([connections containsObject:connection]);
}

void HAPPlatformTCPStreamClose(
        HAPPlatformTCPStreamManagerRef _Nonnull tcpStreamManager,
        HAPPlatformTCPStreamRef tcpStream) {
    HAPPrecondition(tcpStreamManager);
    HAPPrecondition(tcpStream);

    HAPLogDebug(&tcp_log, "Close %lx", tcpStream);

    Connection* connection = (__bridge Connection*) (void*) tcpStream;
    HAPAssert([connections containsObject:connection]);

    [connections removeObject:connection];

    nw_connection_cancel(connection.socket);
}

void HAPPlatformTCPStreamUpdateInterests(
        HAPPlatformTCPStreamManagerRef _Nonnull tcpStreamManager,
        HAPPlatformTCPStreamRef tcpStream,
        HAPPlatformTCPStreamEvent interests,
        HAPPlatformTCPStreamEventCallback _Nullable callback,
        void* _Nullable context) {
    HAPPrecondition(tcpStreamManager);
    HAPPrecondition(tcpStream);

    Connection* connection = (__bridge Connection*) (void*) tcpStream;
    HAPAssert([connections containsObject:connection]);

    connection.interests = interests;
    connection.callback = callback;
    connection.context = context;

    if (interests.hasSpaceAvailable && connection.state == nw_connection_state_ready) {
        dispatch_async(dispatch_get_main_queue(), ^{
            EventCallback(connection, false);
        });
    }
    if (interests.hasBytesAvailable && connection.buffer.length > 0) {
        dispatch_async(dispatch_get_main_queue(), ^{
            EventCallback(connection, true);
        });
    }

    HAPLogDebug(
            &tcp_log,
            "SetInterest %lx: %s%s",
            tcpStream,
            interests.hasBytesAvailable ? "hasBytesAvailable " : "",
            interests.hasSpaceAvailable ? "hasSpaceAvailable " : "");
}

HAP_RESULT_USE_CHECK
HAPError HAPPlatformTCPStreamRead(
        HAPPlatformTCPStreamManagerRef _Nonnull tcpStreamManager,
        HAPPlatformTCPStreamRef tcpStream __unused,
        void* _Nonnull bytes,
        size_t maxBytes,
        size_t* _Nonnull numBytes) {
    HAPPrecondition(tcpStreamManager);
    HAPPrecondition(bytes);
    HAPPrecondition(numBytes);

    Connection* connection = (__bridge Connection*) (void*) tcpStream;
    HAPAssert([connections containsObject:connection]);

    NSData* buffer = connection.buffer;
    if (!buffer.length) {
        *numBytes = 0;
        return kHAPError_None;
    }

    size_t n = buffer.length;
    if (n > maxBytes) {
        n = maxBytes;
    }
    memcpy(bytes, buffer.bytes, *numBytes = n);
    HAPAssert(n <= buffer.length);
    connection.buffer = [buffer subdataWithRange:NSMakeRange(n, buffer.length - n)];

    // HAPLogBufferDebug(&tcp_log, bytes, n, "Read %lx", tcpStream);

    return kHAPError_None;
}

HAP_RESULT_USE_CHECK
HAPError HAPPlatformTCPStreamWrite(
        HAPPlatformTCPStreamManagerRef _Nonnull tcpStreamManager,
        HAPPlatformTCPStreamRef tcpStream,
        const void* _Nonnull bytes,
        size_t maxBytes,
        size_t* _Nonnull numBytes) {
    HAPPrecondition(tcpStreamManager);
    HAPPrecondition(bytes);
    HAPPrecondition(numBytes);

    Connection* connection = (__bridge Connection*) (void*) tcpStream;
    HAPAssert([connections containsObject:connection]);

    // HAPLogBufferDebug(&tcp_log, bytes, maxBytes, "Write %lx", tcpStream);

    nw_content_context_t context = nw_content_context_create("data");
    nw_connection_send(
            connection.socket,
            dispatch_data_create(bytes, maxBytes, dispatch_get_main_queue(), DISPATCH_DATA_DESTRUCTOR_DEFAULT),
            context,
            true,
            ^(nw_error_t error) {
                HAPAssert(!error);
                EventCallback(connection, false);
            });
    *numBytes = maxBytes;

    return kHAPError_None;
}

File: PAL/Mock/HAPPlatformTCPStreamManager.c
// Copyright (c) 2015-2019 The HomeKit ADK Contributors
//
// Licensed under the Apache License, Version 2.0 (the “License”);
// you may not use this file except in compliance with the License.
// See [CONTRIBUTORS.md] for the list of HomeKit ADK project authors.

#include <stdlib.h>

#include "HAPPlatformTCPStreamManager+Init.h"
#include "HAPPlatformTCPStreamManager+Test.h"

static const HAPLogObject logObject = { .subsystem = kHAPPlatform_LogSubsystem, .category = "TCPStreamManager" };

static HAPNetworkPort _port = 1024;

void HAPPlatformTCPStreamManagerCreate(
        HAPPlatformTCPStreamManagerRef tcpStreamManager,
        const HAPPlatformTCPStreamManagerOptions* options) {
    HAPPrecondition(tcpStreamManager);
    HAPPrecondition(options);
    HAPPrecondition(options->tcpStreams);

    HAPRawBufferZero(options->tcpStreams, options->numTCPStreams * sizeof *options->tcpStreams);

    HAPRawBufferZero(tcpStreamManager, sizeof *tcpStreamManager);
    tcpStreamManager->tcpStreams = options->tcpStreams;
    tcpStreamManager->numTCPStreams = options->numTCPStreams;
    tcpStreamManager->numBufferBytes =
            options->numBufferBytes ? options->numBufferBytes : kHAPPlatformTCPStreamManager_NumBufferBytes;

    tcpStreamManager->port = _port++;

    for (size_t i = 0; i < tcpStreamManager->numTCPStreams; i++) {
        HAPPlatformTCPStream* tcpStream = &tcpStreamManager->tcpStreams[i];
        tcpStream->tcpStreamManager = tcpStreamManager;
    }
}

HAP_RESULT_USE_CHECK
HAPNetworkPort HAPPlatformTCPStreamManagerGetListenerPort(HAPPlatformTCPStreamManagerRef tcpStreamManager) {
    HAPPrecondition(tcpStreamManager);

    return tcpStreamManager->port;
}

HAP_RESULT_USE_CHECK
bool HAPPlatformTCPStreamManagerIsListenerOpen(HAPPlatformTCPStreamManagerRef tcpStreamManager) {
    HAPPrecondition(tcpStreamManager);

    return tcpStreamManager->callback != NULL;
}

void HAPPlatformTCPStreamManagerOpenListener(
        HAPPlatformTCPStreamManagerRef tcpStreamManager,
        HAPPlatformTCPStreamListenerCallback callback,
        void* _Nullable context) {
    HAPPrecondition(tcpStreamManager);
    HAPPrecondition(callback);

    HAPLogInfo(&logObject, "%s(%u).", __func__, tcpStreamManager->port);
    HAPPrecondition(!tcpStreamManager->callback);
    HAPPrecondition(!tcpStreamManager->context);
    tcpStreamManager->callback = callback;
    tcpStreamManager->context = context;
}

void HAPPlatformTCPStreamManagerCloseListener(HAPPlatformTCPStreamManagerRef tcpStreamManager) {
    HAPPrecondition(tcpStreamManager);

    HAPLogInfo(&logObject, "%s(%u).", __func__, tcpStreamManager->port);
    tcpStreamManager->callback = NULL;
    tcpStreamManager->context = NULL;
}

HAP_RESULT_USE_CHECK
HAPError HAPPlatformTCPStreamManagerConnectToListener(
        HAPPlatformTCPStreamManagerRef tcpStreamManager,
        HAPPlatformTCPStreamRef* tcpStream) {
    HAPPrecondition(tcpStreamManager);
    HAPPrecondition(HAPPlatformTCPStreamManagerIsListenerOpen(tcpStreamManager));
    HAPPrecondition(tcpStream);

    for (size_t i = 0; i < tcpStreamManager->numTCPStreams; i++) {
        HAPPlatformTCPStream* stream = &tcpStreamManager->tcpStreams[i];
        if (stream->isActive) {
            continue;
        }

        // Open connection.
        HAPLogInfo(&logObject, "Opened connection: %p.", (const void*) stream);
        stream->isActive = true;
        stream->rx.maxBytes = tcpStreamManager->numBufferBytes;
        stream->rx.bytes = calloc(1, stream->rx.maxBytes);
        HAPAssert(stream->rx.bytes);
        stream->tx.maxBytes = tcpStreamManager->numBufferBytes;
        stream->tx.bytes = calloc(1, stream->tx.maxBytes);
        HAPAssert(stream->tx.bytes);
        *tcpStream = (HAPPlatformTCPStreamRef) stream;

        // Inform delegate.
        HAPPrecondition(tcpStreamManager->callback);
        tcpStreamManager->callback(tcpStreamManager, tcpStreamManager->context);

        return kHAPError_None;
    }

    HAPLogError(&logObject, "TCP stream manager cannot accept more connections.");
    return kHAPError_OutOfResources;
}

HAP_RESULT_USE_CHECK
HAPError HAPPlatformTCPStreamManagerAcceptTCPStream(
        HAPPlatformTCPStreamManagerRef tcpStreamManager,
        HAPPlatformTCPStreamRef* tcpStream) {
    HAPPrecondition(tcpStreamManager);
    HAPPrecondition(tcpStream);

    for (size_t i = 0; i < tcpStreamManager->numTCPStreams; i++) {
        HAPPlatformTCPStream* stream = &tcpStreamManager->tcpStreams[i];
        if (!stream->isActive) {
            continue;
        }
        if (stream->isConnected) {
            continue;
        }

        // Open connection.
        HAPLogInfo(&logObject, "Accepted connection: %p.", (const void*) stream);
        stream->isConnected = true;
        *tcpStream = (HAPPlatformTCPStreamRef) stream;

        return kHAPError_None;
    }

    HAPLog(&logObject, "No acceptable connections found.");
    return kHAPError_Unknown;
}

static void Invalidate(HAPPlatformTCPStreamManagerRef tcpStreamManager, HAPPlatformTCPStreamRef tcpStream_) {
    HAPPrecondition(tcpStreamManager);
    HAPPrecondition(tcpStream_);
    HAPPlatformTCPStream* tcpStream = (HAPPlatformTCPStream*) tcpStream_;

    if (tcpStream->invokeCallbackTimer) {
        HAPPlatformTimerDeregister(tcpStream->invokeCallbackTimer);
    }
    free(tcpStream->rx.bytes);
    free(tcpStream->tx.bytes);
    HAPRawBufferZero(tcpStream, sizeof *tcpStream);
}

void HAPPlatformTCPStreamCloseOutput(
        HAPPlatformTCPStreamManagerRef tcpStreamManager,
        HAPPlatformTCPStreamRef tcpStream HAP_UNUSED) {
    HAPPrecondition(tcpStreamManager);

    HAPLogError(&logObject, "[NYI] %s.", __func__);
    HAPFatalError();
}

void HAPPlatformTCPStreamClose(HAPPlatformTCPStreamManagerRef tcpStreamManager, HAPPlatformTCPStreamRef tcpStream_) {
    HAPPrecondition(tcpStreamManager);
    HAPPrecondition(tcpStream_);
    HAPPlatformTCPStream* tcpStream = (HAPPlatformTCPStream*) tcpStream_;
    HAPAssert(tcpStream->isActive);

    HAPAssert(!tcpStream->rx.isClosed);
    tcpStream->rx.isClosed = true;
    tcpStream->tx.isClosed = true;

    // Release resources when both sides closed.
    if (tcpStream->rx.isClosed && tcpStream->rx.isClientClosed) {
        HAPLogDebug(&logObject, "[%p] Closing.", (const void*) tcpStream);
        HAPAssert(tcpStream->tx.isClosed);
        Invalidate(tcpStreamManager, tcpStream_);
    }
}

static void InvokeCallback(HAPPlatformTCPStreamManagerRef tcpStreamManager, HAPPlatformTCPStreamRef tcpStream_) {
    HAPPrecondition(tcpStreamManager);
    HAPPrecondition(tcpStream_);
    HAPPlatformTCPStream* tcpStream = (HAPPlatformTCPStream*) tcpStream_;
    HAPAssert(tcpStream->isActive);

    if (!tcpStream->callback) {
        return;
    }

    HAPPlatformTCPStreamEvent event = { .hasBytesAvailable = tcpStream->rx.numBytes || tcpStream->rx.isClientClosed,
                                        .hasSpaceAvailable = tcpStream->tx.numBytes < tcpStream->tx.maxBytes };

    if (tcpStream->rx.isClosed || !tcpStream->interests.hasBytesAvailable) {
        event.hasBytesAvailable = false;
    }
    if (tcpStream->tx.isClosed || !tcpStream->interests.hasSpaceAvailable) {
        event.hasSpaceAvailable = false;
    }
    if (event.hasBytesAvailable || event.hasSpaceAvailable) {
        tcpStream->callback(tcpStreamManager, tcpStream_, event, tcpStream->context);
    }
}

static void InvokeCallbackExpired(HAPPlatformTimerRef timer, void* _Nullable context) {
    HAPPrecondition(context);
    HAPPlatformTCPStream* tcpStream = context;
    HAPPlatformTCPStreamRef tcpStream_ = (HAPPlatformTCPStreamRef) tcpStream;
    HAPPrecondition(timer == tcpStream->invokeCallbackTimer);
    tcpStream->invokeCallbackTimer = 0;

    HAPPlatformTCPStreamManagerRef tcpStreamManager = tcpStream->tcpStreamManager;
    InvokeCallback(tcpStreamManager, tcpStream_);
}

void HAPPlatformTCPStreamManagerClientClose(
        HAPPlatformTCPStreamManagerRef tcpStreamManager,
        HAPPlatformTCPStreamRef tcpStream_) {
    HAPPrecondition(tcpStreamManager);
    HAPPrecondition(tcpStream_);
    HAPPlatformTCPStream* tcpStream = (HAPPlatformTCPStream*) tcpStream_;
    HAPAssert(tcpStream->isActive);

    HAPAssert(!tcpStream->rx.isClientClosed);
    tcpStream->rx.isClientClosed = true;
    InvokeCallback(tcpStreamManager, tcpStream_);

    // Release resources when both sides closed.
    if (tcpStream->rx.isClosed && tcpStream->rx.isClientClosed) {
        HAPLogDebug(&logObject, "[%p] Closing (Client closed).", (const void*) tcpStream);
        HAPAssert(tcpStream->tx.isClosed);
        Invalidate(tcpStreamManager, tcpStream_);
    }
}

void HAPPlatformTCPStreamUpdateInterests(
        HAPPlatformTCPStreamManagerRef tcpStreamManager,
        HAPPlatformTCPStreamRef tcpStream_,
        HAPPlatformTCPStreamEvent interests,
        HAPPlatformTCPStreamEventCallback _Nullable callback,
        void* _Nullable context) {
    HAPPrecondition(tcpStreamManager);
    HAPPrecondition(tcpStream_);
    HAPPlatformTCPStream* tcpStream = (HAPPlatformTCPStream*) tcpStream_;
    HAPPrecondition(tcpStream->isActive);
    HAPPrecondition(tcpStream->isConnected);
    HAPPrecondition(!(interests.hasBytesAvailable || interests.hasSpaceAvailable) || callback != NULL);

    HAPError err;

    tcpStream->interests = interests;
    tcpStream->callback = callback;
    tcpStream->context = context;

    if (!tcpStream->invokeCallbackTimer) {
        err = HAPPlatformTimerRegister(&tcpStream->invokeCallbackTimer, 0, InvokeCallbackExpired, tcpStream);
        HAPAssert(!err);
    }
}

HAP_RESULT_USE_CHECK
HAPError HAPPlatformTCPStreamRead(
        HAPPlatformTCPStreamManagerRef tcpStreamManager,
        HAPPlatformTCPStreamRef tcpStream_,
        void* bytes,
        size_t maxBytes,
        size_t* numBytes) {
    HAPPrecondition(tcpStreamManager);
    HAPPrecondition(tcpStream_);
    HAPPlatformTCPStream* tcpStream = (HAPPlatformTCPStream*) tcpStream_;
    HAPAssert(tcpStream->isActive);
    HAPPrecondition(bytes);
    HAPPrecondition(numBytes);

    *numBytes = HAPMin(tcpStream->rx.numBytes, maxBytes);
    HAPRawBufferCopyBytes(bytes, HAPNonnullVoid(tcpStream->rx.bytes), *numBytes);
    HAPRawBufferCopyBytes(
            HAPNonnullVoid(tcpStream->rx.bytes),
            &((uint8_t*) tcpStream->rx.bytes)[*numBytes],
            tcpStream->rx.numBytes - *numBytes);
    tcpStream->rx.numBytes -= *numBytes;

    if (!*numBytes && !tcpStream->rx.isClosed) {
        return kHAPError_Busy;
    }
    return kHAPError_None;
}

HAP_RESULT_USE_CHECK
HAPError HAPPlatformTCPStreamClientRead(
        HAPPlatformTCPStreamManagerRef tcpStreamManager,
        HAPPlatformTCPStreamRef tcpStream_,
        void* bytes,
        size_t maxBytes,
        size_t* numBytes) {
    HAPPrecondition(tcpStreamManager);
    HAPPrecondition(tcpStream_);
    HAPPlatformTCPStream* tcpStream = (HAPPlatformTCPStream*) tcpStream_;
    HAPAssert(tcpStream->isActive);
    HAPPrecondition(bytes);
    HAPPrecondition(numBytes);

    *numBytes = 0;
    size_t n;
    InvokeCallback(tcpStreamManager, tcpStream_);
    do {
        n = HAPMin(tcpStream->tx.numBytes, maxBytes - *numBytes);
        HAPRawBufferCopyBytes(&((uint8_t*) bytes)[*numBytes], HAPNonnullVoid(tcpStream->tx.bytes), n);
        HAPRawBufferCopyBytes(
                HAPNonnullVoid(tcpStream->tx.bytes), &((uint8_t*) tcpStream->tx.bytes)[n], tcpStream->tx.numBytes - n);
        tcpStream->tx.numBytes -= n;
        *numBytes += n;
        InvokeCallback(tcpStreamManager, tcpStream_);
    } while (*numBytes < maxBytes && n && !tcpStream->tx.isClosed);
    HAPAssert(*numBytes <= maxBytes);

    if (!*numBytes && !tcpStream->tx.isClosed) {
        return kHAPError_Busy;
    }
    return kHAPError_None;
}

HAP_RESULT_USE_CHECK
HAPError HAPPlatformTCPStreamWrite(
        HAPPlatformTCPStreamManagerRef tcpStreamManager,
        HAPPlatformTCPStreamRef tcpStream_,
        const void* bytes,
        size_t maxBytes,
        size_t* numBytes) {
    HAPPrecondition(tcpStreamManager);
    HAPPrecondition(tcpStream_);
    HAPPlatformTCPStream* tcpStream = (HAPPlatformTCPStream*) tcpStream_;
    HAPAssert(tcpStream->isActive);
    HAPPrecondition(bytes);
    HAPPrecondition(numBytes);

    if (tcpStream->tx.isClosed) {
        return kHAPError_Unknown;
    }

    *numBytes = HAPMin(tcpStream->tx.maxBytes - tcpStream->tx.numBytes, maxBytes);
    HAPRawBufferCopyBytes(&((uint8_t*) tcpStream->tx.bytes)[tcpStream->tx.numBytes], bytes, *numBytes);
    tcpStream->tx.numBytes += *numBytes;

    if (!*numBytes) {
        return kHAPError_Busy;
    }
    return kHAPError_None;
}

HAP_RESULT_USE_CHECK
HAPError HAPPlatformTCPStreamClientWrite(
        HAPPlatformTCPStreamManagerRef tcpStreamManager,
        HAPPlatformTCPStreamRef tcpStream_,
        const void* bytes,
        size_t maxBytes,
        size_t* numBytes) {
    HAPPrecondition(tcpStreamManager);
    HAPPrecondition(tcpStream_);
    HAPPlatformTCPStream* tcpStream = (HAPPlatformTCPStream*) tcpStream_;
    HAPAssert(tcpStream->isActive);
    HAPPrecondition(bytes);
    HAPPrecondition(numBytes);

    *numBytes = 0;
    size_t n;
    InvokeCallback(tcpStreamManager, tcpStream_);
    do {
        n = HAPMin(tcpStream->rx.maxBytes - tcpStream->rx.numBytes, maxBytes - *numBytes);
        HAPRawBufferCopyBytes(
                &((uint8_t*) tcpStream->rx.bytes)[tcpStream->rx.numBytes], &((const uint8_t*) bytes)[*numBytes], n);
        tcpStream->rx.numBytes += n;
        *numBytes += n;
        InvokeCallback(tcpStreamManager, tcpStream_);
    } while (*numBytes < maxBytes && n && !tcpStream->rx.isClosed);
    HAPAssert(*numBytes <= maxBytes);

    if (!*numBytes && !tcpStream->rx.isClosed) {
        return kHAPError_Busy;
    }
    return kHAPError_None;
}
